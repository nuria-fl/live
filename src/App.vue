<template>
  <main>
    <section v-if="!hasStarted" class="Splash">
      <h1>Live</h1>
      <p>A game about survival</p>
      <p>What's your name, survivor?</p>
      <form @submit.prevent="start">
        <input
          v-model="startUsername"
          type="text"
          required
          class="Splash__input"
        />
        <button type="submit" class="Btn">
          New Game
        </button>
      </form>
    </section>
    <main-component v-else @newGame="start" />
    <footer class="Footer">
      <game-status-button v-if="hasStarted" />
      Bug report via
      <a href="https://github.com/nuria-fl/live">Github</a>
      or
      <a href="https://twitter.com/pincfloit">Twitter</a>
      · ☕ <a href="https://ko-fi.com/G2G5AV2X">Buy Me a Coffee</a>
    </footer>
  </main>
</template>

<script>
import { mapState, mapMutations } from 'vuex'
import Main from '@/views/Main'
import GameStatusButton from '@/components/GameStatusButton'
import tracking from '@/utils/tracking'

export default {
  components: {
    MainComponent: Main,
    GameStatusButton
  },
  data() {
    return {
      hasStarted: false,
      startUsername: ''
    }
  },
  computed: {
    ...mapState(['username'])
  },
  created() {
    this.initUsername()
    this.startUsername = this.username
  },
  methods: {
    ...mapMutations(['initUsername', 'setUsername']),
    start() {
      tracking.trackEvent('game', 'start')
      this.setUsername(this.startUsername)
      this.hasStarted = true
    }
  }
}
</script>

<style lang="scss">
body {
  margin: 0;
  font-family: 'Fira Mono', monospace;
  font-size: 0.95em;
  overscroll-behavior: contain;
}
li {
  margin: 1em 0;
}
.Splash {
  padding: 3em 0 4em;
  text-align: center;
  &__input {
    margin-right: 0.5rem;
  }
}
.Footer {
  width: 100%;
  margin-top: 2em;
  padding: 1em;
  background: rgba(255, 255, 255, 0.8);
  @media screen and (min-width: 680px) {
    position: fixed;
    bottom: 0;
    left: 0;
  }
}
</style>
