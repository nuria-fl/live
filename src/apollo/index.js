import ApolloClient from 'apollo-boost'

export const apolloClient = new ApolloClient({
  uri: process.env.VUE_APP_API_URL
})
