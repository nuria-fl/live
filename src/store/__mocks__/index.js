import Vue from 'vue';
import Vuex from 'vuex';
import items from '../../data/items'

Vue.use(Vuex);

export const getters = {
  isInventoryFull: jest.fn().mockReturnValue(false),
  slotsInInventoryLeft: jest.fn().mockReturnValue(8),
  craftableItems: jest.fn().mockReturnValue()
};

export const mutations = {
  pauseGame: jest.fn(),
  playGame: jest.fn(),
  changePage: jest.fn(),
  increaseDayCount: jest.fn(),
  decrease: jest.fn(),
  increase: jest.fn(),
  addInventory: jest.fn(),
  removeInventory: jest.fn(),
  disable: jest.fn(),
  enable: jest.fn(),
  enableFire: jest.fn(),
  disableFire: jest.fn(),
  getSick: jest.fn(),
  getCured: jest.fn()
};

export const actions = {
  initInventory: jest.fn(),
  decreaseAsync: jest.fn(),
  increaseAsync: jest.fn(),
  scavenge: jest.fn(),
  hunt: jest.fn(),
  removeItemsById: jest.fn()
};

export const state = {
  gameOver: false,
  causeOfDeath: null,
  paused: false,
  disabled: false,
  stats: {
    health: 100,
    water: 100,
    food: 100,
    sleep: 100
  },
  existingItems: items,
  inventory: [],
  hasFire: false,
  isSick: false,
  daysSurvived: 0,
  currentPage: 'home'
};

// eslint-disable-next-line no-underscore-dangle
export function __createMocks(custom = { getters: {}, mutations: {}, actions: {}, state: {} }) {
  const mockGetters = Object.assign({}, getters, custom.getters);
  const mockMutations = Object.assign({}, mutations, custom.mutations);
  const mockActions = Object.assign({}, actions, custom.actions);
  const mockState = Object.assign({}, state, custom.state);

  return {
    getters: mockGetters,
    mutations: mockMutations,
    actions: mockActions,
    state: mockState,
    store: new Vuex.Store({
      getters: mockGetters,
      mutations: mockMutations,
      actions: mockActions,
      state: mockState,
    }),
  };
}

export const store = __createMocks().store;
