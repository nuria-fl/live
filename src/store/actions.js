import utils from '@/utils'
import eventBus from '@/utils/eventBus'

export default {
  initInventory({ state, commit }) {
    const scavengeableItems = state.existingItems.filter(
      (item) => item.action === 'scavenge'
    )
    const startingItems = utils.randomizeItems(scavengeableItems, 2)
    startingItems.forEach((item) => commit('addInventory', item))
  },
  decreaseAsync({ state, commit }, { stat, amount, time }) {
    return new Promise((resolve, reject) => {
      commit('disable')
      if (!state.gameOver) {
        setTimeout(function() {
          commit('enable')
          commit('decrease', { stat, amount })
          resolve()
        }, time)
      } else {
        reject(new Error('game is over'))
      }
    })
  },
  increaseAsync({ state, commit }, { stat, amount, time }) {
    return new Promise((resolve, reject) => {
      commit('disable')
      if (!state.gameOver) {
        setTimeout(function() {
          commit('enable')
          commit('increase', { stat, amount })
          resolve()
        }, time)
      } else {
        reject(new Error('game is over'))
      }
    })
  },
  scavenge({ state, getters, commit }, { time }) {
    return new Promise((resolve, reject) => {
      commit('disable')
      if (!state.gameOver) {
        setTimeout(function() {
          const scavengeableItems = state.existingItems.filter(
            (item) => item.action === 'scavenge'
          )

          const numberOfItems =
            getters.slotsInInventoryLeft > 3 ? 3 : getters.slotsInInventoryLeft

          const newItems = utils.randomizeItems(
            scavengeableItems,
            numberOfItems
          )

          newItems.forEach((item) => {
            commit('addInventory', item)
          })

          commit('decrease', { stat: 'energy', amount: 5 })
          commit('decrease', { stat: 'water', amount: 5 })
          commit('decrease', { stat: 'food', amount: 3 })
          commit('enable')

          resolve(newItems)
        }, time)
      } else {
        reject(new Error('game is over'))
      }
    })
  },
  hunt({ state, commit, dispatch }, { time, weapon }) {
    return new Promise((resolve, reject) => {
      commit('disable')
      if (!state.gameOver) {
        setTimeout(function() {
          commit('decrease', { stat: 'energy', amount: 10 })
          commit('decrease', { stat: 'water', amount: 10 })
          commit('decrease', { stat: 'food', amount: 6 })
          commit('enable')

          const isSuccesful = utils.calculateProbability(8)

          if (isSuccesful) {
            dispatch('handleItemDegradation', weapon)
            const huntableItems = state.existingItems.filter(
              (item) => item.action === 'hunt'
            )

            const newItems = utils.randomizeItems(huntableItems, 1)

            newItems.forEach((item) => {
              commit('addInventory', item)
            })

            resolve(newItems)
          } else {
            resolve(false)
          }
        }, time)
      } else {
        reject(new Error('game is over'))
      }
    })
  },
  handleItemDegradation({ commit }, item) {
    if (item.usesUntilBreakdown > 1) {
      commit('degradeItem', item.uid)
    } else {
      commit('removeInventory', item.uid)
      eventBus.$emit('showNotification', {
        text: `${item.name} has broken`
      })
    }
  },
  removeItemsById({ state, commit }, items) {
    const findItemByName = (itemName) => {
      return state.inventory.find((item) => item.id === itemName)
    }

    items.forEach((item) => {
      const itemToRemove = findItemByName(item)
      commit('removeInventory', itemToRemove.uid)
    })
  }
}
