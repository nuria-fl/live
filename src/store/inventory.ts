import { defineStore } from "pinia";

import { MAXINVENTORY } from "../data/constants";
import items, { Item } from "../data/items";
import recipes, { Recipe } from "../data/recipes";
import upgrades, { Upgrade } from "../data/upgrades";
import utils from "../utils";
import { eventBus } from "../utils/eventBus";

const findIndexById = (uid: string, collection: Item[]) => {
	return collection.findIndex((item) => item.uid === uid);
};

const isCraftable = (
	inventoryState: Item[],
	hasFire: boolean,
	recipe: Recipe | Upgrade,
) => {
	const inventory = inventoryState.map((item) => item.id);
	const currentItems: string[] = [];

	let isCraftable = false;

	recipe.itemsNeeded.forEach((itemNeeded) => {
		const idx = inventory.indexOf(itemNeeded);

		if (idx !== -1) {
			// if we find the item we remove it from the temp inventory
			inventory.splice(idx, 1);
			currentItems.push(itemNeeded);
		} else {
			isCraftable = false;
		}
	});

	// TODO: is length enough? deepEqual?
	const hasItems = currentItems.length === recipe.itemsNeeded.length;

	let hasTools = true;

	recipe.toolsNeeded.forEach((tool) => {
		const idx = inventory.indexOf(tool);

		if (idx === -1) {
			hasTools = false;
		} else {
			hasTools = true;
		}
	});

	if (hasItems && hasTools) {
		isCraftable = true;
		if (
			"upgradesNeeded" in recipe &&
			recipe.upgradesNeeded.indexOf("fire") > -1 &&
			!hasFire
		) {
			isCraftable = false;
		}
	}

	return isCraftable;
};

export const useInventoryStore = defineStore("inventory", {
	state: (): {
		existingItems: Item[];
		inventory: Item[];
		hasFire: boolean;
	} => ({
		existingItems: items,
		inventory: [],
		hasFire: false,
	}),
	getters: {
		slotsInInventoryLeft(state) {
			return MAXINVENTORY - state.inventory.length;
		},
		isInventoryFull(state) {
			return state.inventory.length === MAXINVENTORY;
		},
		recipes(state) {
			return recipes.map((recipe) => {
				return {
					...recipe,
					isCraftable: isCraftable(state.inventory, state.hasFire, recipe),
				};
			});
		},
		upgrades(state) {
			return upgrades.map((upgrade) => {
				return {
					...upgrade,
					isCraftable: isCraftable(state.inventory, state.hasFire, upgrade),
				};
			});
		},
	},
	actions: {
		initInventory() {
			const scavengeableItems = this.existingItems.filter(
				(item) => item.action === "scavenge",
			);
			const startingItems = utils.randomizeItems(scavengeableItems, 2);
			startingItems.forEach((item) => {
				this.addInventory(item);
			});
		},
		degradeItem(uid: string) {
			const idx = findIndexById(uid, this.inventory);

			if (idx !== -1) {
				this.inventory[idx].usesUntilBreakdown--;
			}
		},
		addInventory(item: Item) {
			if (this.inventory.length < MAXINVENTORY) {
				item.uid = utils.generateId();
				this.inventory.push(item);
			}
		},
		removeInventory(uid: string) {
			const idx = findIndexById(uid, this.inventory);

			if (idx !== -1) {
				this.inventory.splice(idx, 1);
			}
		},
		handleItemDegradation(item: Item) {
			if (item.usesUntilBreakdown > 1) {
				this.degradeItem(item.uid!);
			} else {
				this.removeInventory(item.uid!);
				eventBus.$emit("showNotification", {
					text: `${item.name} has broken`,
				});
			}
		},
		removeItemsById(items: string[]) {
			const findItemByName = (itemName: string) => {
				return this.inventory.find((item) => item.id === itemName);
			};

			items.forEach((itemName) => {
				const itemToRemove = findItemByName(itemName);
				if (!itemToRemove) {
					throw new Error(`Item ${itemName} not found`);
				}
				this.removeInventory(itemToRemove.uid!);
			});
		},
		enableFire() {
			this.hasFire = true;
		},
		disableFire() {
			this.hasFire = false;
		},
	},
});
