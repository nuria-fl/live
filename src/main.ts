import "./assets/styles/index.scss";

import { createApp } from "vue";

import App from "./App.vue";
import { createAppStore } from "./store";

const store = createAppStore();

const app = createApp(App);

app.use(store);

app.mount("#app");
