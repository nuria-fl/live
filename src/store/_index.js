import { createStore } from "vuex";

import actions from "./_actions";
import mutations from "./_mutations";
import { getters, state } from "./_state";

const store = createStore({
	state,
	getters,
	mutations,
	actions,
});

export default store;
