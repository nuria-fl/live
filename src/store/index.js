import { createStore } from 'vuex'
import { state, getters } from './state'
import mutations from './mutations'
import actions from './actions'

const store = createStore({
  state,
  getters,
  mutations,
  actions
})

export default store
