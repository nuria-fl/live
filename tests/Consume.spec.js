import Vuex from 'vuex'
import { shallow, createLocalVue } from 'vue-test-utils'
import { __createMocks as createStoreMocks } from '../src/store'

import Consume from '../src/components/Consume'

// Tell Jest to use the mock implementation of the store
jest.mock('../src/store')

const localVue = createLocalVue()

localVue.use(Vuex)

// TODO: test sick/cure

describe('Consume', () => {
  let storeMocks
  let wrapper

  beforeEach(() => {
    // Create a fresh store and wrapper instance for every test case.
    storeMocks = createStoreMocks()
    wrapper = shallow(Consume, {
      store: storeMocks.store,
      localVue,
      propsData: {
        item: {
          type: 'food',
          id: 'berries',
          name: 'Berries',
          value: 3,
          risk: 0,
          action: 'scavenge',
          description: 'Small amount of vitamins and fiber'
        }
      }
    })
  })

  test('It should match snapshot', () => {
    expect(wrapper.vm.$el).toMatchSnapshot()
  })

  test('Calls removeInventory when discarting', () => {
    wrapper.find('.Btn:first-child').trigger('click')
    expect(storeMocks.mutations.removeInventory).toBeCalled()
  })
  test('Calls removeInventory when consuming', () => {
    wrapper.find('.Btn:last-child').trigger('click')
    expect(storeMocks.mutations.removeInventory).toBeCalled()
  })
  test('Calls increase when consuming', () => {
    wrapper.find('.Btn:last-child').trigger('click')
    expect(storeMocks.mutations.increase).toBeCalled()
  })
})
