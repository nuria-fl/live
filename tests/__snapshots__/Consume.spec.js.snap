// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Consume It should match snapshot 1`] = `
<article
  class="Item"
>
  <h4>
    
    Berries 
    
    <span
      class="Item__stats"
    >
      
      +3 food
    
    </span>
  </h4>
   
  <p>
    Small amount of vitamins and fiber
  </p>
   
  <div
    class="Item__actions Item__actions--multi"
  >
    <button
      class="Btn"
    >
      
      discard
    
    </button>
    <button
      class="Btn"
    >
      
      consume
    
    </button>
  </div>
</article>
`;
